src/data/sum/interval.lean

```diff
@@ -0,0 +1,160 @@
+
+/-- Lifts maps `α₁ → β₁ → finset γ₁` and `α₂ → β₂ → finset γ₂` to a map
+`α₁ ⊕ α₂ → β₁ ⊕ β₂ → finset (γ₁ ⊕ γ₂)`. Could be generalized to alternative monads if we can make
+sure to keep computability and universe polymorphism. -/
+def sum_lift₂ : Π (a : α₁ ⊕ α₂) (b : β₁ ⊕ β₂), finset (γ₁ ⊕ γ₂)
+| (inl a) (inl b) := (f a b).map embedding.inl
+| (inl a) (inr b) := ∅
+| (inr a) (inl b) := ∅
+| (inr a) (inr b) := (g a b).map embedding.inr
+
+@[simp] lemma sum_lift₂_inl_inl (a : α₁) (b : β₁) :
+  sum_lift₂ f g (inl a) (inl b) = (f a b).map embedding.inl := rfl
+
+@[simp] lemma sum_lift₂_inl_inr (a : α₁) (b : β₂) : sum_lift₂ f g (inl a) (inr b) = ∅ := rfl
+@[simp] lemma sum_lift₂_inr_inl (a : α₂) (b : β₁) : sum_lift₂ f g (inr a) (inl b) = ∅ := rfl
+
+@[simp] lemma sum_lift₂_inr_inr (a : α₂) (b : β₂) :
+  sum_lift₂ f g (inr a) (inr b) = (g a b).map embedding.inr := rfl
+
+variables {f f₁ g₁ g f₂ g₂} {a : α₁ ⊕ α₂} {b : β₁ ⊕ β₂} {c : γ₁ ⊕ γ₂}
+
+lemma mem_sum_lift₂ :
+  c ∈ sum_lift₂ f g a b ↔ (∃ a₁ b₁ c₁, a = inl a₁ ∧ b = inl b₁ ∧ c = inl c₁ ∧ c₁ ∈ f a₁ b₁)
+    ∨ ∃ a₂ b₂ c₂, a = inr a₂ ∧ b = inr b₂ ∧ c = inr c₂ ∧ c₂ ∈ g a₂ b₂ :=
+begin
+  split,
+  { cases a; cases b,
+    { rw [sum_lift₂, mem_map],
+      rintro ⟨c, hc, rfl⟩,
+      exact or.inl ⟨a, b, c, rfl, rfl, rfl, hc⟩ },
+    { refine λ h, (not_mem_empty _ h).elim },
+    { refine λ h, (not_mem_empty _ h).elim },
+    { rw [sum_lift₂, mem_map],
+      rintro ⟨c, hc, rfl⟩,
+      exact or.inr ⟨a, b, c, rfl, rfl, rfl, hc⟩ } },
+  { rintro (⟨a, b, c, rfl, rfl, rfl, h⟩ | ⟨a, b, c, rfl, rfl, rfl, h⟩); exact mem_map_of_mem _ h }
+end
+
+lemma inl_mem_sum_lift₂ {c₁ : γ₁} :
+  inl c₁ ∈ sum_lift₂ f g a b ↔ ∃ a₁ b₁, a = inl a₁ ∧ b = inl b₁ ∧ c₁ ∈ f a₁ b₁ :=
+begin
+  rw [mem_sum_lift₂, or_iff_left],
+  simp only [exists_and_distrib_left, exists_eq_left'],
+  rintro ⟨_, _, c₂, _, _, h, _⟩,
+  exact inl_ne_inr h,
+end
+
+lemma inr_mem_sum_lift₂ {c₂ : γ₂} :
+  inr c₂ ∈ sum_lift₂ f g a b ↔ ∃ a₂ b₂, a = inr a₂ ∧ b = inr b₂ ∧ c₂ ∈ g a₂ b₂ :=
+begin
+  rw [mem_sum_lift₂, or_iff_right],
+  simp only [exists_and_distrib_left, exists_eq_left'],
+  rintro ⟨_, _, c₂, _, _, h, _⟩,
+  exact inr_ne_inl h,
+end
+
+lemma sum_lift₂_nonempty :
+  (sum_lift₂ f g a b).nonempty ↔ (∃ a₁ b₁, a = inl a₁ ∧ b = inl b₁ ∧ (f a₁ b₁).nonempty)
+    ∨ ∃ a₂ b₂, a = inr a₂ ∧ b = inr b₂ ∧ (g a₂ b₂).nonempty :=
+begin
+  split,
+  { rintro ⟨c | c, hc⟩,
+    { rw inl_mem_sum_lift₂ at hc,
+      obtain ⟨a, b, ha, hb, hc⟩ := hc,
+      exact or.inl ⟨a, b, ha, hb, c, hc⟩ },
+    { rw inr_mem_sum_lift₂ at hc,
+      obtain ⟨a, b, ha, hb, hc⟩ := hc,
+      exact or.inr ⟨a, b, ha, hb, c, hc⟩ } },
+  { rintro (⟨a, b, rfl, rfl, h⟩ | ⟨a, b, rfl, rfl, h⟩); exact h.map }
+end
+
+lemma sum_lift₂_eq_empty :
+  (sum_lift₂ f g a b) = ∅ ↔ (∀ a₁ b₁, a = inl a₁ → b = inl b₁ → f a₁ b₁ = ∅)
+    ∧ ∀ a₂ b₂, a = inr a₂ → b = inr b₂ → g a₂ b₂ = ∅ :=
+begin
+  refine ⟨λ h, ⟨_, _⟩, λ h, _⟩,
+  any_goals { rintro a b rfl rfl, exact map_eq_empty.1 h },
```

Comment:
I don't really like this use of `any_goals`: at this point there are three goals and this will work on the first two. How do you feel about
```
  refine ⟨λ h, _, λ h, _⟩,
  { split; { rintro a b rfl rfl, exact map_eq_empty.1 h } },
```
instead? It more clearly shows that exactly the first two goals should be closed, and has no chance of making progress on the later goals; in my view this is easier to maintain.

Reply:
Sure, this is as many lines.